crossbeam-epoch/src/internal.rs

```diff
@@ -389,38 +387,47 @@ impl Local {
-                let current = Epoch::starting();
-                let previous = self
-                    .epoch
-                    .compare_and_swap(current, new_epoch, Ordering::SeqCst);
-                debug_assert_eq!(current, previous, "participant was expected to be unpinned");
-                // We add a compiler fence to make it less likely for LLVM to do something wrong
-                // here.  Formally, this is not enough to get rid of data races; practically,
-                // it should go a long way.
-                atomic::compiler_fence(Ordering::SeqCst);
-            } else {
-                self.epoch.store(new_epoch, Ordering::Relaxed);
-                atomic::fence(Ordering::SeqCst);
+            // Now we must store the global epoch into `self.epoch` and execute a `SeqCst` fence.
+            // The fence makes sure that any future loads from `Atomic`s will not happen before this
+            // store.
+            let mut current = Epoch::starting();
+            let mut new = self.global().epoch.load(Ordering::Relaxed).pinned();
+
+            loop {
+                if cfg!(any(target_arch = "x86", target_arch = "x86_64")) {
+                    // HACK(stjepang): On x86 architectures there are two different ways of
+                    // executing a `SeqCst` fence.
+                    //
+                    // 1. `atomic::fence(SeqCst)`, which compiles into a `mfence` instruction.
+                    // 2. `_.compare_and_swap(_, _, SeqCst)`, which compiles into a `lock cmpxchg`
+                    //    instruction.
+                    //
+                    // Both instructions have the effect of a full barrier, but benchmarks have
+                    // shown that the second one makes pinning faster in this particular case.  It
+                    // is not clear that this is permitted by the C++ memory model (SC fences work
+                    // very differently from SC accesses), but experimental evidence suggests that
+                    // this works fine.  Using inline assembly would be a viable (and correct)
+                    // alternative, but alas, that is not possible on stable Rust.
+                    let previous = self.epoch.compare_and_swap(current, new, Ordering::SeqCst);
+                    debug_assert_eq!(current, previous, "participant was expected to be unpinned");
+
+                    // We add a compiler fence to make it less likely for LLVM to do something wrong
+                    // here.  Formally, this is not enough to get rid of data races; practically, it
+                    // should go a long way.
+                    atomic::compiler_fence(Ordering::SeqCst);
+                } else {
+                    self.epoch.store(new, Ordering::Relaxed);
+                    atomic::fence(Ordering::SeqCst);
+                }
+
+                // Now we validate that the value we read from the global epoch is not stale.
```

Comment:
What is the motivation for validation phase? I'm afraid that this loop might slow things down a bit.

Could you run `cargo bench` and see if this affects performance?

Reply:
It's for helping the global epoch advancement in general. It doesn't seem to have a meaningful effect on the performance.

* master:
```
test multi_alloc_defer_free  ... bench:   4,135,110 ns/iter (+/- 566,246)
test multi_defer             ... bench:   2,154,671 ns/iter (+/- 212,984)
test single_alloc_defer_free ... bench:          43 ns/iter (+/- 0)
test single_defer            ... bench:          16 ns/iter (+/- 0)

test multi_flush  ... bench:  22,928,416 ns/iter (+/- 1,546,218)
test single_flush ... bench:         136 ns/iter (+/- 1)

test multi_pin  ... bench:   4,179,789 ns/iter (+/- 167,156)
test single_pin ... bench:           5 ns/iter (+/- 0)
```

* free-after-3
```
test multi_alloc_defer_free  ... bench:   4,176,331 ns/iter (+/- 513,890)
test multi_defer             ... bench:   2,179,475 ns/iter (+/- 197,665)
test single_alloc_defer_free ... bench:          43 ns/iter (+/- 1)
test single_defer            ... bench:          16 ns/iter (+/- 0)

test multi_flush  ... bench:  21,997,370 ns/iter (+/- 3,973,497)
test single_flush ... bench:         135 ns/iter (+/- 0)

test multi_pin  ... bench:   4,173,010 ns/iter (+/- 389,623)
test single_pin ... bench:           5 ns/iter (+/- 0)

```
* free-after-3 without validation
```
test multi_alloc_defer_free  ... bench:   4,140,581 ns/iter (+/- 426,444)
test multi_defer             ... bench:   2,160,064 ns/iter (+/- 196,742)
test single_alloc_defer_free ... bench:          43 ns/iter (+/- 0)
test single_defer            ... bench:          16 ns/iter (+/- 0)

test multi_flush  ... bench:  22,800,624 ns/iter (+/- 2,119,408)
test single_flush ... bench:         136 ns/iter (+/- 1)

test multi_pin  ... bench:   4,228,682 ns/iter (+/- 380,561)
test single_pin ... bench:           5 ns/iter (+/- 0)
```